File: js/src/translation/emit-common.js
// State shared by the emitters: target names for every declaration, the
// source-to-target mapping, and the encodings and assumptions a translation
// relies on. Every emitter records what it did here, so the translation
// contract lists exactly the choices that were made for this program.

import { unsupported } from './diagnostics.js';
import { typeKey } from './types.js';

/** Assumption texts are fixed so that contracts compare across runtimes. */
export const ASSUMPTIONS = {
  nonAborting: {
    id: 'non-aborting-executions',
    statement: 'the translation agrees with the source on executions that do not abort; the source aborts on machine-integer overflow, checked conversion failure, division by zero or an explicit panic, and the target computes an unspecified value there instead',
  },
};

export class EmitState {
  /**
   * @param {object} program checked portable-core program
   * @param {string} language target language
   * @param {(name: string) => string} ident legal target identifier for a source name
   * @param {Set<string>} reserved names no declaration may take
   * @param {object} options `{ ctorStyle: 'module' | 'data', typeName, valueName, ctorName, moduleSegment, generated, typeSpace, modulesShareTermSpace, modulesShareTypeSpace }`
   */
  constructor(program, language, ident, reserved, options = {}) {
    this.program = program;
    this.language = language;
    this.ident = ident;
    this.reserved = reserved;
    this.options = options;
    this.names = new Map();
    this.ctorNames = new Map();
    this.moduleNames = new Map();
    this.taken = new Map();
    this.mappings = [];
    this.assumptions = new Map();
    this.encodings = new Map();
    this.theorems = [];
    this.assign();
  }

  /** Target names are fixed up front, so references never depend on emission order. */
  assign() {
    const namespace = (key) => {
      if (!this.taken.has(key)) this.taken.set(key, new Set(this.reserved));
      return this.taken.get(key);
    };
    const claim = (space, name) => {
      const used = namespace(space);
      let candidate = name;
      for (let index = 2; used.has(candidate) || this.isGenerated(candidate, used); index += 1) candidate = `${name}_${index}`;
      used.add(candidate);
      for (const extra of this.generatedFor(candidate)) used.add(extra);
      return candidate;
    };
    const moduleName = this.options.moduleSegment ?? this.ident;
    // JavaScript modules are values; Rust modules share the type namespace.
    let moduleSpace = 'module';
    if (this.options.modulesShareTermSpace) moduleSpace = 'term';
    else if (this.options.modulesShareTypeSpace) moduleSpace = 'type';
    for (const entry of this.program.declarations.values()) {
      const segments = entry.modulePath.map((segment, index) => {
        const key = entry.modulePath.slice(0, index + 1).join('.');
        if (!this.moduleNames.has(key)) {
          const parent = entry.modulePath.slice(0, index).join('.');
          this.moduleNames.set(key, claim(`${moduleSpace}:${parent}`, moduleName(segment)));
        }
        return this.moduleNames.get(key);
      });
      const space = `term:${entry.modulePath.join('.')}`;
      const base = entry.k === 'data'
        ? (this.options.typeName ?? this.ident)(entry.name)
        : (this.options.valueName ?? this.ident)(entry.name);
      const local = claim(entry.k === 'data' && this.options.typeSpace ? `type:${entry.modulePath.join('.')}` : space, base);
      this.names.set(entry.fullName, { local, path: segments });
      if (entry.k === 'data') {
        for (const ctor of entry.ctors) {
          const ctorBase = (this.options.ctorName ?? this.ident)(ctor.name);
          const ctorSpace = this.options.ctorStyle === 'module' ? space : `ctor:${entry.fullName}`;
          this.ctorNames.set(`${entry.fullName}#${ctor.name}`, claim(ctorSpace, ctorBase));
        }
      }
    }
  }

  /**
   * Names a local binder may not take: a local named like a module, a
   * top-level function or a constructor would shadow it in the target.
   */
  localReserved() {
    const names = new Set(this.reserved);
    for (const used of this.taken.values()) for (const name of used) names.add(name);
    return names;
  }

  isGenerated(candidate, used) {
    return this.generatedFor(candidate).some((name) => used.has(name));
  }

  /** Names the target itself derives from a declaration (Rocq's `f_equation`, `T_ind`). */
  generatedFor(name) {
    return (this.options.generated ?? (() => []))(name);
  }

  localName(fullName) {
    return this.names.get(fullName).local;
  }

  modulePath(fullName) {
    return this.names.get(fullName).path;
  }

  moduleName(sourcePath) {
    return this.moduleNames.get(sourcePath.join('.'));
  }

  /** Fully qualified reference from the top level. */
  ref(fullName, separator = '.') {
    const { local, path } = this.names.get(fullName);
    return [...path, local].join(separator);
  }

  ctorLocal(dataName, ctor) {
    return this.ctorNames.get(`${dataName}#${ctor}`);
  }

  ctorRef(dataName, ctor, separator = '.') {
    const local = this.ctorLocal(dataName, ctor);
    if (this.options.ctorStyle === 'module') return [...this.modulePath(dataName), local].join(separator);
    return `${this.ref(dataName, separator)}${separator}${local}`;
  }

  map(entry, target) {
    this.mappings.push({
      kind: entry.k === 'fn' ? 'function' : entry.k,
      source: entry.fullName,
      target: [...this.modulePath(entry.fullName), target].join('.'),
      sourceSpan: entry.span ? { start: entry.span.start, end: entry.span.end } : null,
    });
    if (entry.k === 'data') {
      for (const ctor of entry.ctors) {
        this.mappings.push({
          kind: 'constructor',
          source: `${entry.fullName}.${ctor.name}`,
          target: this.ctorRef(entry.fullName, ctor.name),
          sourceSpan: entry.span ? { start: entry.span.start, end: entry.span.end } : null,
        });
      }
    }
  }

  assume(assumption, detail) {
    if (!this.assumptions.has(assumption.id)) this.assumptions.set(assumption.id, { ...assumption, details: [] });
    const record = this.assumptions.get(assumption.id);
    if (detail && !record.details.includes(detail)) record.details.push(detail);
  }

  fixedToUnbounded(type) {
    const key = typeKey(type);
    this.encode(`machine-integer:${key}`, `${key} values are represented by the target's unbounded ${type.signed ? 'integers' : 'naturals'}; they agree while no operation overflows`);
    this.assume(ASSUMPTIONS.nonAborting, `${key} arithmetic stays in range`);
  }

  checkedToTotal(operation) {
    this.assume(ASSUMPTIONS.nonAborting, `checked ${operation} does not fail`);
  }

  abortToTotal(message) {
    this.assume(ASSUMPTIONS.nonAborting, `no abort: ${message}`);
  }

  encode(id, statement) {
    if (!this.encodings.has(id)) this.encodings.set(id, { id, statement });
  }

  theorem(entry, name, details = {}) {
    this.theorems.push({ source: entry.fullName, target: [...this.modulePath(entry.fullName), name].join('.'), kind: 'theorem', ...details });
  }

  assertionTheorem(name, effect) {
    this.theorems.push({
      source: `main:assert@${effect.span ? effect.span.start : '?'}`,
      target: name,
      kind: 'assertion',
      closedGoal: true,
    });
  }

  assumptionList() {
    return [...this.assumptions.values()];
  }

  encodingList() {
    return [...this.encodings.values()];
  }
}

/** Declarations every declaration refers to (types, functions, constructors, lemmas). */
export function dependencies(entry) {
  const found = new Set();
  const types = (type) => {
    if (type?.kind === 'data') found.add(type.name);
  };
  const visit = (node) => {
    if (!node || typeof node !== 'object') return;
    if (Array.isArray(node)) {
      node.forEach(visit);
      return;
    }
    if (node.kind === 'data') found.add(node.name);
    if (node.k === 'call') found.add(node.fn);
    if (node.k === 'ctor') found.add(node.data);
    for (const value of Object.values(node)) visit(value);
  };
  if (entry.k === 'data') entry.ctors.forEach((ctor) => ctor.fields.forEach((field) => types(field.type)));
  if (entry.k === 'fn') {
    entry.params.forEach((param) => types(param.type));
    types(entry.ret);
    visit(entry.body);
  }
  if (entry.k === 'theorem') {
    entry.binders.forEach((binder) => types(binder.type));
    visit(entry.prop);
    const lemmas = (plan) => {
      if (plan.k === 'close') {
        plan.hints.lemmas.forEach((name) => found.add(name));
        plan.hints.unfold.forEach((name) => found.add(name));
      } else plan.cases.forEach((kase) => lemmas(kase.plan));
    };
    lemmas(entry.proof.plan);
  }
  found.delete(entry.fullName);
  return found;
}

/**
 * Declarations in an order where each follows everything it uses. Source
 * order is kept where possible and, among ready declarations, the ones in
 * the module being emitted come first so modules stay contiguous.
 */
export function orderDeclarations(program, { requireContiguousModules = false } = {}) {
  const entries = [...program.declarations.values()];
  const position = new Map(entries.map((entry, index) => [entry.fullName, index]));
  const pending = new Map(entries.map((entry) => [entry.fullName, dependencies(entry)]));
  const done = new Set();
  const order = [];
  let current = [];
  while (order.length < entries.length) {
    const ready = entries.filter((entry) => !done.has(entry.fullName)
      && [...pending.get(entry.fullName)].every((name) => done.has(name) || !position.has(name)));
    if (!ready.length) {
      const stuck = entries.find((entry) => !done.has(entry.fullName));
      throw unsupported('cyclic declarations', `${stuck.fullName} depends on declarations that depend on it, which needs mutual definitions`, stuck.span);
    }
    const shared = (entry) => {
      let common = 0;
      while (common < current.length && common < entry.modulePath.length && current[common] === entry.modulePath[common]) common += 1;
      return common;
    };
    ready.sort((left, right) => (shared(right) - shared(left))
      || (position.get(left.fullName) - position.get(right.fullName)));
    const [next] = ready;
    order.push(next);
    done.add(next.fullName);
    current = next.modulePath;
  }
  if (requireContiguousModules) {
    const closed = new Set();
    let open = [];
    for (const entry of order) {
      for (let depth = 0; depth < open.length; depth += 1) {
        if (entry.modulePath[depth] !== open[depth]) {
          for (let index = depth; index < open.length; index += 1) closed.add(open.slice(0, index + 1).join('.'));
          break;
        }
      }
      for (let depth = 1; depth <= entry.modulePath.length; depth += 1) {
        const key = entry.modulePath.slice(0, depth).join('.');
        if (closed.has(key)) {
          throw unsupported('interleaved modules', `module ${key} would have to be reopened after its declarations depend on later ones; the target cannot reopen a module`, entry.span);
        }
      }
      open = entry.modulePath;
    }
  }
  return order;
}

File: rust/src/translation/emit_common.rs
//! State shared by the emitters.
//!
//! The state holds target names for every declaration, the source-to-target
//! mapping, and the encodings and assumptions a translation relies on. Every
//! emitter records what it did here, so the translation contract lists exactly
//! the choices that were made for this program.
//!
//! Mirrors `js/src/translation/emit-common.js`.

use std::collections::{BTreeSet, HashMap, HashSet};

use serde::{Deserialize, Serialize};

use super::diagnostics::{unsupported, Result};
use super::ir::{Decl, Effect, Expr, Node, Pattern, Plan, Program, Prop};
use super::types::Type;
use super::{Language, Span};

/// A fixed assumption text, so that contracts compare across runtimes.
#[derive(Clone, Copy, Debug)]
pub struct AssumptionText {
    pub id: &'static str,
    pub statement: &'static str,
}

pub const NON_ABORTING: AssumptionText = AssumptionText {
    id: "non-aborting-executions",
    statement: "the translation agrees with the source on executions that do not abort; the source aborts on machine-integer overflow, checked conversion failure, division by zero or an explicit panic, and the target computes an unspecified value there instead",
};

/// Where a source declaration went in the target.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Mapping {
    pub kind: String,
    pub source: String,
    pub target: String,
    pub source_span: Option<Span>,
}

/// An assumption the translation relies on, with the program-specific details.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Assumption {
    pub id: String,
    pub statement: String,
    pub details: Vec<String>,
}

/// A representation choice the translation made.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Encoding {
    pub id: String,
    pub statement: String,
}

/// A theorem or assertion the target states, and how it is discharged.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TheoremRecord {
    pub source: String,
    pub target: String,
    pub kind: String,
    pub closed_goal: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub discharge: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub check: Option<String>,
}

/// An emitted target program together with its translation contract.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Emitted {
    pub language: Language,
    pub text: String,
    pub mappings: Vec<Mapping>,
    pub assumptions: Vec<Assumption>,
    pub encodings: Vec<Encoding>,
    pub theorems: Vec<TheoremRecord>,
    pub entry: Option<String>,
}

/// How constructors are named: inside a module next to their type (Rocq), or under the type.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum CtorStyle {
    Module,
    #[default]
    Data,
}

/// Target naming rules for [`EmitState`].
#[derive(Clone, Copy, Default)]
#[allow(clippy::struct_excessive_bools)] // the flags mirror independent target namespaces
pub struct EmitOptions {
    pub ctor_style: CtorStyle,
    pub type_name: Option<fn(&str) -> String>,
    pub value_name: Option<fn(&str) -> String>,
    pub ctor_name: Option<fn(&str) -> String>,
    pub module_segment: Option<fn(&str) -> String>,
    /// Names the target itself derives from a declaration (Rocq's `f_equation`, `T_ind`).
    pub generated: Option<fn(&str) -> Vec<String>>,
    pub type_space: bool,
    pub modules_share_term_space: bool,
    pub modules_share_type_space: bool,
    /// Names root declarations may not take because the target's prelude has them.
    pub root_reserved: &'static [&'static str],
}

struct Target {
    local: String,
    path: Vec<String>,
}

pub struct EmitState<'p> {
    pub program: &'p Program,
    pub language: Language,
    pub ident: fn(&str) -> String,
    reserved: BTreeSet<String>,
    options: EmitOptions,
    names: HashMap<String, Target>,
    ctor_names: HashMap<String, String>,
    module_names: HashMap<String, String>,
    taken: HashMap<String, HashSet<String>>,
    claimed: BTreeSet<String>,
    pub mappings: Vec<Mapping>,
    assumptions: Vec<Assumption>,
    encodings: Vec<Encoding>,
    pub theorems: Vec<TheoremRecord>,
}

impl<'p> EmitState<'p> {
    #[must_use]
    pub fn new(
        program: &'p Program,
        language: Language,
        ident: fn(&str) -> String,
        reserved: &[&str],
        options: EmitOptions,
    ) -> Self {
        let mut state = Self {
            program,
            language,
            ident,
            reserved: reserved.iter().map(|name| (*name).to_owned()).collect(),
            options,
            names: HashMap::new(),
            ctor_names: HashMap::new(),
            module_names: HashMap::new(),
            taken: HashMap::new(),
            claimed: BTreeSet::new(),
            mappings: Vec::new(),
            assumptions: Vec::new(),
            encodings: Vec::new(),
            theorems: Vec::new(),
        };
        state.assign();
        state
    }

    /// Target names are fixed up front, so references never depend on emission order.
    fn assign(&mut self) {
        let module_name = self.options.module_segment.unwrap_or(self.ident);
        // JavaScript modules are values; Rust modules share the type namespace.
        let module_space = if self.options.modules_share_term_space {
            "term"
        } else if self.options.modules_share_type_space {
            "type"
        } else {
            "module"
        };
        let program = self.program;
        for entry in &program.declarations {
            let module_path = entry.module_path();
            let mut segments = Vec::new();
            for (index, segment) in module_path.iter().enumerate() {
                let key = module_path[..=index].join(".");
                if !self.module_names.contains_key(&key) {
                    let parent = module_path[..index].join(".");
                    let name =
                        self.claim(&format!("{module_space}:{parent}"), &module_name(segment));
                    self.module_names.insert(key.clone(), name);
                }
                segments.push(self.module_names[&key].clone());
            }
            let space = format!("term:{}", module_path.join("."));
            let base = match entry {
                Decl::Data(_) => self.options.type_name.unwrap_or(self.ident)(entry.name()),
                _ => self.options.value_name.unwrap_or(self.ident)(entry.name()),
            };
            let local_space = if matches!(entry, Decl::Data(_)) && self.options.type_space {
                format!("type:{}", module_path.join("."))
            } else {
                space.clone()
            };
            let local = self.claim(&local_space, &base);
            self.names.insert(
                entry.full_name().to_owned(),
                Target {
                    local,
                    path: segments,
                },
            );
            if let Decl::Data(data) = entry {
                for ctor in &data.ctors {
                    let ctor_base = self.options.ctor_name.unwrap_or(self.ident)(&ctor.name);
                    let ctor_space = if self.options.ctor_style == CtorStyle::Module {
                        space.clone()
                    } else {
                        format!("ctor:{}", data.full_name)
                    };
                    let name = self.claim(&ctor_space, &ctor_base);
                    self.ctor_names
                        .insert(format!("{}#{}", data.full_name, ctor.name), name);
                }
            }
        }
    }

    fn claim(&mut self, space: &str, name: &str) -> String {
        if !self.taken.contains_key(space) {
            // Root declarations may not redeclare the target's prelude (`root_reserved`).
            let mut used: HashSet<String> = self.reserved.iter().cloned().collect();
            if space.ends_with(':') {
                used.extend(
                    self.options
                        .root_reserved
                        .iter()
                        .map(|name| (*name).to_owned()),
                );
            }
            self.taken.insert(space.to_owned(), used);
        }
        let generated = self.options.generated;
        let used = self
            .taken
            .get_mut(space)
            .expect("namespace was just created");
        let is_generated = |candidate: &str, used: &HashSet<String>| {
            generated.is_some_and(|generated| {
                generated(candidate).iter().any(|name| used.contains(name))
            })
        };
        let mut candidate = name.to_owned();
        let mut index = 2;
        while used.contains(&candidate) || is_generated(&candidate, used) {
            candidate = format!("{name}_{index}");
            index += 1;
        }
        let mut names = vec![candidate.clone()];
        if let Some(generated) = generated {
            names.extend(generated(&candidate));
        }
        for name in names {
            used.insert(name.clone());
            self.claimed.insert(name);
        }
        candidate
    }

    /// Names a local binder may not take: a local named like a module, a
    /// top-level function or a constructor would shadow it in the target.
    #[must_use]
    pub fn local_reserved(&self) -> BTreeSet<String> {
        self.reserved.union(&self.claimed).cloned().collect()
    }

    /// The target name of a declaration within its module.
    ///
    /// # Panics
    /// When `full_name` is not a declaration of the program.
    #[must_use]
    pub fn local_name(&self, full_name: &str) -> &str {
        &self.names[full_name].local
    }

    /// The target module path of a declaration.
    ///
    /// # Panics
    /// When `full_name` is not a declaration of the program.
    #[must_use]
    pub fn module_path(&self, full_name: &str) -> &[String] {
        &self.names[full_name].path
    }

    /// The target name of a source module.
    #[must_use]
    pub fn module_name(&self, source_path: &[String]) -> Option<&str> {
        self.module_names
            .get(&source_path.join("."))
            .map(String::as_str)
    }

    /// Fully qualified reference from the top level.
    #[must_use]
    pub fn reference(&self, full_name: &str, separator: &str) -> String {
        let target = &self.names[full_name];
        let mut parts = target.path.clone();
        parts.push(target.local.clone());
        parts.join(separator)
    }

    /// The target name of a constructor within its namespace.
    ///
    /// # Panics
    /// When the constructor is not declared.
    #[must_use]
    pub fn ctor_local(&self, data_name: &str, ctor: &str) -> &str {
        &self.ctor_names[&format!("{data_name}#{ctor}")]
    }

    /// Fully qualified constructor reference from the top level.
    #[must_use]
    pub fn ctor_ref(&self, data_name: &str, ctor: &str, separator: &str) -> String {
        let local = self.ctor_local(data_name, ctor);
        if self.options.ctor_style == CtorStyle::Module {
            let mut parts = self.module_path(data_name).to_vec();
            parts.push(local.to_owned());
            return parts.join(separator);
        }
        format!("{}{separator}{local}", self.reference(data_name, separator))
    }

    /// Records where a declaration (and a data type's constructors) went.
    pub fn map(&mut self, entry: &Decl, target: &str) {
        let source_span = entry.span();
        let mut path = self.module_path(entry.full_name()).to_vec();
        path.push(target.to_owned());
        self.mappings.push(Mapping {
            kind: match entry {
                Decl::Fn(_) => "function",
                other => other.kind(),
            }
            .to_owned(),
            source: entry.full_name().to_owned(),
            target: path.join("."),
            source_span,
        });
        if let Decl::Data(data) = entry {
            for ctor in &data.ctors {
                self.mappings.push(Mapping {
                    kind: "constructor".to_owned(),
                    source: format!("{}.{}", data.full_name, ctor.name),
                    target: self.ctor_ref(&data.full_name, &ctor.name, "."),
                    source_span,
                });
            }
        }
    }

    pub fn assume(&mut self, assumption: AssumptionText, detail: Option<&str>) {
        let index = if let Some(index) = self
            .assumptions
            .iter()
            .position(|record| record.id == assumption.id)
        {
            index
        } else {
            self.assumptions.push(Assumption {
                id: assumption.id.to_owned(),
                statement: assumption.statement.to_owned(),
                details: Vec::new(),
            });
            self.assumptions.len() - 1
        };
        let record = &mut self.assumptions[index];
        if let Some(detail) = detail {
            if !record.details.iter().any(|known| known == detail) {
                record.details.push(detail.to_owned());
            }
        }
    }

    pub fn fixed_to_unbounded(&mut self, ty: &Type) {
        let key = ty.key();
        let unbounded = if matches!(ty, Type::Fixed { signed: true, .. }) {
            "integers"
        } else {
            "naturals"
        };
        self.encode(
            &format!("machine-integer:{key}"),
            &format!("{key} values are represented by the target's unbounded {unbounded}; they agree while no operation overflows"),
        );
        self.assume(
            NON_ABORTING,
            Some(&format!("{key} arithmetic stays in range")),
        );
    }

    pub fn checked_to_total(&mut self, operation: &str) {
        self.assume(
            NON_ABORTING,
            Some(&format!("checked {operation} does not fail")),
        );
    }

    pub fn abort_to_total(&mut self, message: &str) {
        self.assume(NON_ABORTING, Some(&format!("no abort: {message}")));
    }

    pub fn encode(&mut self, id: &str, statement: &str) {
        if !self.encodings.iter().any(|encoding| encoding.id == id) {
            self.encodings.push(Encoding {
                id: id.to_owned(),
                statement: statement.to_owned(),
            });
        }
    }

    /// Records a theorem; `bounded` theorems are checked by the source
    /// kernel and tested on bounded inputs in the target.
    pub fn theorem(&mut self, full_name: &str, name: &str, closed_goal: bool, bounded: bool) {
        let mut path = self.module_path(full_name).to_vec();
        path.push(name.to_owned());
        self.theorems.push(TheoremRecord {
            source: full_name.to_owned(),
            target: path.join("."),
            kind: "theorem".to_owned(),
            closed_goal,
            discharge: bounded.then(|| "source-kernel".to_owned()),
            check: bounded.then(|| "bounded".to_owned()),
        });
    }

    pub fn assertion_theorem(&mut self, name: &str, effect: &Effect) {
        let span = match effect {
            Effect::Print { span, .. } | Effect::Let { span, .. } | Effect::Assert { span, .. } => {
                *span
            }
        };
        self.theorems.push(TheoremRecord {
            source: format!(
                "main:assert@{}",
                span.map_or_else(|| "?".to_owned(), |span| span.start.to_string())
            ),
            target: name.to_owned(),
            kind: "assertion".to_owned(),
            closed_goal: true,
            discharge: None,
            check: None,
        });
    }

    #[must_use]
    pub fn assumption_list(&self) -> Vec<Assumption> {
        self.assumptions.clone()
    }

    #[must_use]
    pub fn encoding_list(&self) -> Vec<Encoding> {
        self.encodings.clone()
    }

    /// The emitted result with this state's contract.
    #[must_use]
    pub fn finish(&self, text: String, entry: Option<String>) -> Emitted {
        Emitted {
            language: self.language,
            text,
            mappings: self.mappings.clone(),
            assumptions: self.assumption_list(),
            encodings: self.encoding_list(),
            theorems: self.theorems.clone(),
            entry,
        }
    }
}

fn type_dependencies(ty: &Type, found: &mut HashSet<String>) {
    if let Type::Data { name } = ty {
        found.insert(name.clone());
    }
}

fn expr_dependencies(expr: &Expr, found: &mut HashSet<String>) {
    type_dependencies(&expr.ty, found);
    match &expr.node {
        Node::Call { func, .. } => {
            found.insert(func.clone());
        }
        Node::Ctor { data, .. } => {
            found.insert(data.clone());
        }
        Node::Binary {
            domain: Some(domain),
            ..
        } => type_dependencies(domain, found),
        Node::Cast { from, to, .. } => {
            type_dependencies(from, found);
            type_dependencies(to, found);
        }
        Node::Match { cases, .. } => {
            for kase in cases {
                if let Pattern::Ctor { data, .. } = &kase.pattern {
                    found.insert(data.clone());
                }
            }
        }
        _ => {}
    }
    for child in expr.children() {
        expr_dependencies(child, found);
    }
}

fn prop_dependencies(prop: &Prop, found: &mut HashSet<String>) {
    match prop {
        Prop::Forall { binders, body } => {
            for binder in binders {
                type_dependencies(&binder.ty, found);
            }
            prop_dependencies(body, found);
        }
        Prop::And { left, right } | Prop::Or { left, right } | Prop::Implies { left, right } => {
            prop_dependencies(left, found);
            prop_dependencies(right, found);
        }
        Prop::Not { arg } => prop_dependencies(arg, found),
        Prop::Bool { expr } => expr_dependencies(expr, found),
        other => {
            if let Some((_, comparison)) = other.comparison() {
                expr_dependencies(&comparison.left, found);
                expr_dependencies(&comparison.right, found);
                type_dependencies(&comparison.domain, found);
            }
        }
    }
}

/// Declarations every declaration refers to (types, functions, constructors, lemmas).
#[must_use]
pub fn dependencies(entry: &Decl) -> HashSet<String> {
    fn lemmas(plan: &Plan, found: &mut HashSet<String>) {
        match plan {
            Plan::Close { hints } => {
                found.extend(hints.lemmas.iter().cloned());
                found.extend(hints.unfold.iter().cloned());
            }
            Plan::Induction(split) | Plan::Cases(split) => {
                for kase in &split.cases {
                    lemmas(&kase.plan, found);
                }
            }
        }
    }
    let mut found = HashSet::new();
    match entry {
        Decl::Data(data) => {
            for ctor in &data.ctors {
                for field in &ctor.fields {
                    type_dependencies(&field.ty, &mut found);
                }
            }
        }
        Decl::Fn(function) => {
            for param in &function.params {
                type_dependencies(&param.ty, &mut found);
            }
            type_dependencies(&function.ret, &mut found);
            expr_dependencies(&function.body, &mut found);
        }
        Decl::Theorem(theorem) => {
            for binder in &theorem.binders {
                type_dependencies(&binder.ty, &mut found);
            }
            prop_dependencies(&theorem.prop, &mut found);
            lemmas(&theorem.proof.plan, &mut found);
        }
    }
    found.remove(entry.full_name());
    found
}

/// Declarations in an order where each follows everything it uses.
///
/// Source order is kept where possible and, among ready declarations, the
/// ones in the module being emitted come first so modules stay contiguous.
///
/// # Errors
/// On cyclic declarations, and on modules that would have to be reopened
/// when `require_contiguous_modules` is set.
pub fn order_declarations(
    program: &Program,
    require_contiguous_modules: bool,
) -> Result<Vec<&Decl>> {
    let entries = &program.declarations;
    let position: HashMap<&str, usize> = entries
        .iter()
        .enumerate()
        .map(|(index, entry)| (entry.full_name(), index))
        .collect();
    let pending: Vec<HashSet<String>> = entries.iter().map(dependencies).collect();
    let mut done: HashSet<&str> = HashSet::new();
    let mut order: Vec<&Decl> = Vec::new();
    let mut current: &[String] = &[];
    while order.len() < entries.len() {
        let mut ready: Vec<usize> = (0..entries.len())
            .filter(|&index| {
                !done.contains(entries[index].full_name())
                    && pending[index].iter().all(|name| {
                        done.contains(name.as_str()) || !position.contains_key(name.as_str())
                    })
            })
            .collect();
        if ready.is_empty() {
            let stuck = entries
                .iter()
                .find(|entry| !done.contains(entry.full_name()))
                .expect("some declaration is pending");
            return Err(unsupported(
                "cyclic declarations",
                &format!(
                    "{} depends on declarations that depend on it, which needs mutual definitions",
                    stuck.full_name()
                ),
                stuck.span(),
            ));
        }
        let shared = |index: usize| {
            let path = entries[index].module_path();
            current
                .iter()
                .zip(path)
                .take_while(|(left, right)| left == right)
                .count()
        };
        ready.sort_by(|&left, &right| shared(right).cmp(&shared(left)).then(left.cmp(&right)));
        let next = &entries[ready[0]];
        order.push(next);
        done.insert(next.full_name());
        current = next.module_path();
    }
    if require_contiguous_modules {
        let mut closed: HashSet<String> = HashSet::new();
        let mut open: &[String] = &[];
        for entry in &order {
            let path = entry.module_path();
            for depth in 0..open.len() {
                if path.get(depth) != Some(&open[depth]) {
                    for index in depth..open.len() {
                        closed.insert(open[..=index].join("."));
                    }
                    break;
                }
            }
            for depth in 1..=path.len() {
                let key = path[..depth].join(".");
                if closed.contains(&key) {
                    return Err(unsupported(
                        "interleaved modules",
                        &format!("module {key} would have to be reopened after its declarations depend on later ones; the target cannot reopen a module"),
                        entry.span(),
                    ));
                }
            }
            open = path;
        }
    }
    Ok(order)
}
